File: backend/app/schemas/analytics.py
"""Pydantic v2 schemas for the analytics aggregation response."""

from __future__ import annotations

from decimal import Decimal

from pydantic import BaseModel


class PeriodKpis(BaseModel):
    income: Decimal
    expense: Decimal
    cashback: Decimal
    savings: Decimal
    savings_rate: float


class TopCategory(BaseModel):
    category: str
    amount: Decimal
    percent: float


class BudgetAdherence(BaseModel):
    category: str
    amount: Decimal
    spent: Decimal
    remaining: Decimal
    percent_used: float
    status: str


class CardCashback(BaseModel):
    card_id: int
    card_name: str
    spend: Decimal
    cashback_earned: Decimal
    effective_rate: float
    cap: Decimal | None
    cap_used: Decimal | None
    cap_remaining: Decimal | None


class CashbackOptimization(BaseModel):
    total_cashback: Decimal
    by_card: list[CardCashback]


class Growth(BaseModel):
    income: float
    expense: float
    savings: float
    cashback: float


class DateRange(BaseModel):
    start: str
    end: str


class AnalyticsResponse(BaseModel):
    period: str
    range: DateRange
    kpis: PeriodKpis
    previous: PeriodKpis
    growth: Growth
    top_categories: list[TopCategory]
    budget_adherence: list[BudgetAdherence]
    cashback_optimization: CashbackOptimization

File: frontend/src/types/analytics.ts
export type AnalyticsPeriod = 'monthly' | 'quarterly' | 'yearly' | 'financial_year';

export interface PeriodKpis {
  income: string;
  expense: string;
  cashback: string;
  savings: string;
  savings_rate: number;
}

export interface TopCategory {
  category: string;
  amount: string;
  percent: number;
}

export interface BudgetAdherence {
  category: string;
  amount: string;
  spent: string;
  remaining: string;
  percent_used: number;
  status: 'green' | 'yellow' | 'red';
}

export interface CardCashback {
  card_id: number;
  card_name: string;
  spend: string;
  cashback_earned: string;
  effective_rate: number;
  cap: string | null;
  cap_used: string | null;
  cap_remaining: string | null;
}

export interface CashbackOptimization {
  total_cashback: string;
  by_card: CardCashback[];
}

export interface Growth {
  income: number;
  expense: number;
  savings: number;
  cashback: number;
}

export interface DateRange {
  start: string;
  end: string;
}

export interface AnalyticsResponse {
  period: AnalyticsPeriod;
  range: DateRange;
  kpis: PeriodKpis;
  previous: PeriodKpis;
  growth: Growth;
  top_categories: TopCategory[];
  budget_adherence: BudgetAdherence[];
  cashback_optimization: CashbackOptimization;
}
